### PR TITLE
Correctly run the file extension migration on Windows

### PR DESCRIPTION
**Pull request by lukasbableck**
On MariaDB servers on Windows the query returns results even if no uppercase letters are present in the extension column because the comparison is case insensitive apparently. By adding the BINARY keyword the query works correctly no matter the OS

Fixes https://community.contao.org/de/showthread.php?86614-Datenbank-%C3%BCberpr%C3%BCfen-h%C3%A4ngt-bei-quot-FileExtensionMigration-quot

image

### CODE REVIEW

**Comment by ausi** [comment]
We should test if this also works with MySQL 8, as their docs say: _“In MySQL 8.0.22 and later, use of a binary string with any of the MySQL regular expression functions is rejected with ER_CHARACTER_SET_MISMATCH.”_

**Comment by lukasbableck** [comment]
Good point, didn't think of that... Casting the extension column to binary might work
`SELECT * FROM tl_files WHERE CAST(extension AS BINARY) REGEXP BINARY '[[:upper:]]'`

image
image

**Comment by fritzmg** [comment]
The [MySQL 8.0 docs](https://dev.mysql.com/doc/refman/8.0/en/pattern-matching.html) state the following:

> To force a regular expression comparison to be case-sensitive, use a case-sensitive collation, or use the BINARY keyword to make one of the strings a binary string, or specify the c match-control character.

So usage of the `BINARY` keyword should be fine in MySQL 8.0 too? Although it's not clear from the 8.0 docs whether this only applies to `REGEXP_LIKE` (which is only available in 8.0+).

**Comment by lukasbableck** [comment]
So I did some testing now, it definitely doesn't work if you only add the BINARY keyword on MySQL 8.0.

image

Casting does work and produces the result we need.

image

This was tested on MySQL 8.0.33. MariaDB 10 behaves exactly the same with the cast.

File: core-bundle/src/Migration/Version413/FileExtensionMigration.php [hunk 1]
```diff
@@ -36,7 +36,7 @@ public function shouldRun(): bool
             return false;
         }
 
-        return false !== $this->connection->fetchOne("SELECT * FROM tl_files WHERE extension REGEXP '[[:upper:]]' LIMIT 1");
+        return false !== $this->connection->fetchOne("SELECT * FROM tl_files WHERE CAST(extension AS BINARY) REGEXP BINARY '[[:upper:]]' LIMIT 1");
```
Comment (fritzmg) [inline core-bundle/src/Migration/Version413/FileExtensionMigration.php:39]: Shouldn't we apply the same change in line 44?

**Review by ausi — APPROVED** [review]

**Review by fritzmg — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @lukasbableck.

**Pull request merged**